[WIP] Proposed revision to sample_posterior_predictive()

rpgoldman opened the pull request:
Most other sampling functions take varnames, rather than var objects. Extend the set of parameters to accept varnames as an alternative to vars, preserving backwards compatibility.
Also revise the docstring, to clarify the return type and add type comments.

lucianopaz commented on pymc3/sampling.py:
```diff
@@ -1027,8 +1028,11 @@ def stop_tuning(step):
+                                varnames=None,
+                                size=None,
                                 random_seed=None, progressbar=True):
+    # type: (MultiTrace, Optional[int], Optional[Model], Optional[List[RV]], Optional[List[str]], Optional[int], Optional[Any], bool) -> Dict[str, np.ndarray]:
```
No need to comment this out. Type hints are ok because they're supported by python 3.5. What we don't support is variable annotation.
However, the line is too long. Try splitting it with the `black` package. Don't `black` the entire file, just the `sample_posterior_predictive` function.

rpgoldman replied:
Actually, this type hint functions despite being commented out. I'm not sure what the alternative to this form of type annotation for pre-3.6 Python.
So can I use this form in 3.5?

``` python
def sample_posterior_predictive(trace, 
                                                    samples: Optional[int]=None, 
                                                    model: Optional[Model]=None,
                                                    vars: Optional[Iterable[Tensor]]=None,
                                                    varnames: Optional[Iterable[str]]=None,
                                                    size: Optional[int]=None,
                                                    random_seed=None,
                                                    progressbar: bool=True) -> Dict[str, np.ndarray]:
```
I'm afraid I don't understand what is and isn't permitted in 3.5, since I went straight from 2.7 to 3.6.
[Nor, obviously, do I understand what the heck Markdown will do with my whitespace!]

lucianopaz replied:
Type hints are specified in [pep484](https://www.python.org/dev/peps/pep-0484/). How to annotate callables is [here](https://www.python.org/dev/peps/pep-0484/#callable). What you suggest seems to be variable annotation instead

rpgoldman replied:
PEP484 contains this example, which is of the same form as what I gave in my response:

```
def dilate(v: Iterable[Tuple[T, T]], scale: T) -> Iterable[Tuple[T, T]]:
```

I have *also* type hinted local variables, but I moved all that to comments. So I think the example above should be ok with 3.5, if it supports PEP484.

lucianopaz replied:
Ah, great!

lucianopaz commented on pymc3/sampling.py:
```diff
@@ -1043,6 +1047,9 @@ def sample_posterior_predictive(trace, samples=None, model=None, vars=None, size
     vars : iterable
```
You should add a note saying that it's deprecated.

rpgoldman replied:
Oh, are you suggesting I should deprecate this? I wasn't going to go that far -- just add `varnames` as an easier-to-remember alternative. OK, will do.

lucianopaz replied:
I think that supporting two parametrizations is just redundant and I think that we should opt for one or the other. @twiecki, @junpenglao, what do you think?

rpgoldman replied:
I figured you would prefer I not break old code, including example notebooks, that use the old `vars` parameter.
I think `varnames` is better because it preserves the abstraction around the model object, but I could be wrong.

lucianopaz commented on pymc3/sampling.py:
```diff
@@ -1070,6 +1077,10 @@ def sample_posterior_predictive(trace, samples=None, model=None, vars=None, size
+        if vars is not None:
+            raise Exception("Should not specify both vars and varnames arguments.")
+        vars = (model[x] for x in varnames)
     if vars is None:
```
Add an `else` condition that says that `vars` is deprecated in favor of `varnames`

rpgoldman replied:
Will do.

lucianopaz replied:
This can become an else clause

lucianopaz reviewed (changes requested):
Overall, I like that this unifies the API consistently. IMHO, I would have left `vars` and deprecated `varnames`. However, I would like to hear what others think about this. Maybe `varnames` is safer because you ensure that the tensors and RVs are part of the model, whereas, `vars`could hold anything out of the model's context.

Finally, if this goes through, you should add a line at the end of the deprecations section of release-notes.

lucianopaz commented on pymc3/sampling.py:
```diff
@@ -1,3 +1,4 @@
+from typing import Dict, List, Any, Optional
```
Also remove unused imports to fix lint

rpgoldman replied:
I'm not sure what to do about this -- these imports **are** used by mypy, AFAIK.  I don't know how the checker finds type definitions without imports.
I'm happy to do what's right here, but once again, I don't understand how 3.5 functions.

lucianopaz replied:
Travis' logs say

> ++++pylint --rcfile=.pylintrc pymc3
> ************* Module pymc3.sampling
> pymc3/sampling.py:1:
[W0611(unused-import), ] Unused Optional imported from typing

rpgoldman replied:
@lucianopaz At least for now I have 'fixed' this by telling pylint to ignore those imports.  I haven't been able to find an alternative solution, because I don't know otherwise how mypy is to identify the types.  I think what we have now is an acceptable expedient, unless someone can let me know of a better solution. I'll look for some advice myself.

lucianopaz commented:
I think that this change is nice but I think we should discuss two things:

1. Whether we want to change the function parameters to make the API more consistent.
2. Whether to add a very small test environment for python 3.5, so we ensure that we build and at least sample correctly in that version of python. After #3465 and the variables annotations problems, I think that we need this.

@twiecki, @junpenglao, @ericmjl, @ColCarroll

rpgoldman commented:
The deprecation warning was too aggressive.  Fixing that now.

lucianopaz commented on pymc3/sampling.py:
```diff
@@ -1070,6 +1083,12 @@ def sample_posterior_predictive(trace, samples=None, model=None, vars=None, size
+            raise Exception("Should not specify both vars and varnames arguments.")
+        vars = (model[x] for x in varnames)
+    else:
+        raise DeprecationWarning("vars argument is deprecated in favor of varnames.")
```
Use the [warnings package](https://docs.python.org/3.6/library/warnings.html) instead of raising the `DeprecationWarning`. Also, this warning should be raised if `varnames is None` and `vars is not None`.

rpgoldman replied:
Got this now, I believe.  Waiting for the tests to complete.
There's enough uninteresting history on this topic branch that it's probably worth squash-merging it, or I could clean up the history.  Need to rebase it, anyway.

fonnesbeck commented:
Its not clear (to me) that these need to be unified. Sampling takes place in a model context, so it makes sense to pass the PyMC variables directly. Whereas, plotting occurs outside the model context, so the variables are not generally available. To me, there is a clear line between the two, because they are associated with distinct tasks in different phases of the modeling workflow.

rpgoldman commented:
> Its not clear (to me) that these need to be unified. Sampling takes place in a model context, so it makes sense to pass the PyMC variables directly. Whereas, plotting occurs outside the model context, so the variables are not generally available. To me, there is a clear line between the two, because they are associated with distinct tasks in different phases of the modeling workflow.

That may be true, but in addition to being confusing, and having the potential for error that @lucianopaz points out, there's also the issue that since the user is typically working with variable names, they just end up having to write code that looks like:

```
vars=(model[x] for x in name_list)
```
...at least that has been my experience using this API (more accurately, I do that *after* I have encountered the wrong argument error! 😜).  This is inconvenient and potentially error-prone.

Typically I write a function that builds the model and I don't "hang onto" the variable objects, and it's easier to work with the variable names.

lucianopaz commented on pymc3/sampling.py:
```diff
@@ -24,6 +27,8 @@
 import sys
 sys.setrecursionlimit(10000)
 
+assert Any # ugly way to make pyflakes ignore "unused" Any import
```
Better to just not import Any. It's only used in the commented variable annotation

rpgoldman replied:
I have found a way to import this symbol only during type checking, which seems to make pylint and mypy *both* happy.
Rebased and force-pushed this change for review.

lucianopaz commented on pymc3/sampling.py:
```diff
@@ -1,3 +1,5 @@
+from typing import Dict, List, Any, Optional # pylint: disable=unused-import
```
If you remove any from this import, does pylint complain? We prefer to not ignore lint

lucianopaz commented:
@fonnesbeck, sampling from the posterior predictive should also be done within the model's context. Once we get the sample, we can plot outside the context. I'm also hesitant of making this change so I labeled it for discussion.

I now think that using names is safer because that way, we ensure that the variables are defined in the model. If we accept tensors, the users could mix-up tensors that come from two separate models, and could make a strange mess.

junpenglao commented:
I agree with @lucianopaz and this sounds like a good idea to me in general. However, we might want to separate the type hint part with the actual PR? Or is it the plan now to do this for each PR?

You should also do the same change for `sample_posterior_predictive_w`
https://github.com/pymc-devs/pymc3/blob/master/pymc3/sampling.py#L1114

rpgoldman commented:
> I agree with @lucianopaz and this sounds like a good idea to me in general. However, we might want to separate the type hint part with the actual PR? Or is it the plan now to do this for each PR?

I see your point about the type hint part, but separating out the type hints will be quite burdensome (it's not something `diff` is good at).  Perhaps someone who has a better diff tool could split out the type hints as part of squash-merging this PR?

I use type hints both to provide valuable checking, and as a tool in figuring out PyMC3, which is a very complex code base; they have become an integral part of my Python coding process.

> You should also do the same change for `sample_posterior_predictive_w`

But this function doesn't have a `vars` argument at all, so I'm not sure what you would like to have done to it.  I'm happy to make whatever changes, but I need some more direction.

junpenglao commented:
oh right sample_posterior_predictive_w only works for observed variable - never mind :-)

rpgoldman committed:
Revise sample_prior/posterior_predictive() parameters.

Add the option to take varnames (`var_names`), rather than var objects
as parameters.

Extend the set of parameters to accept varnames as an alternative to
vars, preserving backwards compatibility.

Also revise the docstring, to clarify the return type and add type comments.

rpgoldman commented:
Integrated the history, with blind alleys, into a single commit.
Retained the history with tag.

lucianopaz commented on pymc3/sampling.py:
```diff
@@ -529,7 +534,7 @@ def _sample_population(draws, chain, chains, start, random_seed, step, tune,
             trace=None, tune=None, model=None, **kwargs):
     skip_first = kwargs.get('skip_first', 0)
-    refresh_every = kwargs.get('refresh_every', 100)
+    # refresh_every = kwargs.get('refresh_every', 100)
```
Why is this commented out?

rpgoldman replied:
It's commented out because the variable is set but never referenced, and pylint complained. Arguably it should not be in this commit, but it seems too trivial to deserve one of its own.  I could put it back if you like.

I wondered if this was actually trying to set a global that I didn't know about, but I didn't find any explanation -- maybe it's obsolete code that should go, or I suppose I might have missed something important.  If I'm wrong, and it *is* important, we should probably try to make a test that establishes its importance.

lucianopaz replied:
Ah, ok! I think this might have been left dangling after we removed live plots. Could you delete it?

rpgoldman replied:
Done and pushed.

lucianopaz commented on pymc3/sampling.py:
```diff
@@ -1,3 +1,7 @@
+from typing import Dict, List, Optional, TYPE_CHECKING
+if TYPE_CHECKING:
```
Nice

rpgoldman replied:
😀

lucianopaz reviewed (commented):
You still need to add this change to the bottom of the deprecations section of the release notes

lucianopaz commented:
Once #3470 gets sorted out and merged, you'll have to rebase this PR to get the minimum syntax check of python3.5 working

rpgoldman commented:
> Once #3470 gets sorted out and merged, you'll have to rebase this PR to get the minimum syntax check of python3.5 working

I subscribed to that PR. When I hear that it's merged, I will rebase and someone can merge this and close it out.

rpgoldman commented:
> You still need to add this change to the bottom of the deprecations section of the release notes

Just did that and pushed.

lucianopaz reviewed (approved):
The PR looks fine now and could be merged. Do most of us agree that it's better to use `varnames` vs `vars`? I am in favor of `varnames` but I would like know the rest of us approved this change. @twiecki, @fonnesbeck, @junpenglao, @ColCarroll, and anyone else care to give their final opinion?

twiecki commented:
I approve, it's more consistent. But didn't we switched to var_names?

On Sat, May 11, 2019, 07:50 Luciano Paz <notifications@github.com> wrote:

> *@lucianopaz* approved this pull request.
>
> The PR looks fine now and could be merged. Do most of us agree that it's
> better to use varnames vs vars? I am in favor of varnames but I would
> like know the rest of us approved this change. @twiecki
> <https://github.com/twiecki>, @fonnesbeck <https://github.com/fonnesbeck>,
> @junpenglao <https://github.com/junpenglao>, @ColCarroll
> <https://github.com/ColCarroll>, and anyone else care to give their final
> opinion?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pymc-devs/pymc3/pull/3468#pullrequestreview-236363678>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAFETGBIA6KPCQGHXJJOM3TPUZNBJANCNFSM4HLCBZAA>
> .
>

lucianopaz commented:
For now it's still mostly inconsistent but we made the switch to `var_names` in the plots. We can do it here too. Also, I just checked `sample_prior_predictive` and that uses `vars` too. It should be dealt with just like in `sample_posterior_predictive`

twiecki commented:
Yeah let's move everything to var_names.

On Sat, May 11, 2019, 08:12 Luciano Paz <notifications@github.com> wrote:

> For now it's still mostly inconsistent but we made the switch to var_names
> in the plots. We can do it here too. Also, I just checked
> sample_prior_predictive and that uses vars too. It should be dealt with
> just like in sample_posterior_predictive
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pymc-devs/pymc3/pull/3468#issuecomment-491483441>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAFETGBIZQRYRYBUJQNT4HLPUZPWVANCNFSM4HLCBZAA>
> .
>

rpgoldman committed:
Add type hint to modelcontext.

rpgoldman commented:
> For now it's still mostly inconsistent but we made the switch to `var_names` in the plots. We can do it here too. Also, I just checked `sample_prior_predictive` and that uses `vars` too. It should be dealt with just like in `sample_posterior_predictive`

I have pushed the changes to sample_prior_predictive().  I am waiting to make sure they pass the tests. If so, I will do the `varnames` -> `var_names` change, and then I hope we can do the merge.

rpgoldman commented:
OK, I think that this is all ready to go now.  `varnames` have become `var_names`.  If the tests all pass, I will make sure it's properly rebased and squash it down to a single commit for merging, unless there are any further issues?

ColCarroll commented on pymc3/model.py:
```diff
@@ -3,6 +3,7 @@
 import itertools
 import threading
 import warnings
+from typing import Optional, Dict, Any
```
only `Optional` used here?

rpgoldman replied:
See 087ea2d8ebfaa2e34474eba0ce79f3420b78a4b4

ColCarroll commented on pymc3/sampling.py:
```diff
@@ -1070,6 +1083,13 @@ def sample_posterior_predictive(trace, samples=None, model=None, vars=None, size
 
+    if var_names is not None:
+        if vars is not None:
+            raise Exception("Should not specify both vars and var_names arguments.")
```
`ValueError` for a bad argument here (`Exception` is too broad, I think)

rpgoldman replied:
See 5e546b252ec20b5b202037de99f377da91b3356c

ColCarroll commented on pymc3/sampling.py:
```diff
@@ -1273,8 +1303,16 @@ def sample_prior_predictive(samples=500, model=None, vars=None, random_seed=None
+        warnings.warn("vars argument is deprecated in favor of var_names.",
+                      DeprecationWarning)
+    else:
+        raise Exception("Cannot supply both vars and var_names arguments.")
```
`ValueError` here, too

rpgoldman replied:
See 5e546b252ec20b5b202037de99f377da91b3356c

ColCarroll commented on pymc3/sampling.py:
```diff
@@ -1283,8 +1321,9 @@ def sample_prior_predictive(samples=500, model=None, vars=None, random_seed=None
+    assert data is not None
 
-    prior = {}
+    prior: Dict[str, np.ndarray] = {}
```
does this one break on python 3.5? I thought we could only do signature type checks...

rpgoldman replied:
Oh, probably. Sorry. I'll fix that.

rpgoldman replied:
Done.  See 99c7362d81bc729e0fa62de7039bb19b2a8d991d

ColCarroll reviewed (commented):
looks generally good! can you tighten up the exceptions before merge,though?

rpgoldman commented:
@ColCarroll @lucianopaz @twiecki I think I've fixed everything requested.  If the tests go fine, and you like the current version, I will squash this down (or you can do a squash merge), and I hope it will be done.

ColCarroll commented on pymc3/sampling.py:
```diff
@@ -1283,8 +1322,9 @@ def sample_prior_predictive(samples=500, model=None, vars=None, random_seed=None
     values = draw_values([model[name] for name in names], size=samples)
 
     data = {k: v for k, v in zip(names, values)}
+    assert data is not None
```
raw `assert`s are considered bad form (they can be ignored if you use some flag that no one actually uses) -- `if data is None: raise AssertionError`

rpgoldman replied:
Fixed in 65439179b1abe7d8e14b2dbf01d257034f669e7e

rpgoldman replied:
@ColCarroll --
OK, I have no idea why this happened: when I replaced that line per above I now get a bunch of test failures involving failure to find `scipy.misc.factorial`  that look like this:

```
pymc3/tests/test_distributions.py:500: in check_dlogp
    assert_almost_equal(dlogp(pt), ndlogp(pt), decimal=decimals, err_msg=str(pt))
../../../miniconda3/envs/testenv/lib/python3.6/site-packages/numdifftools/core.py:723: in __call__
    *args, **kwds).squeeze()
../../../miniconda3/envs/testenv/lib/python3.6/site-packages/numdifftools/core.py:666: in __call__
    return super(Jacobian, self).__call__(np.atleast_1d(x), *args, **kwds)
../../../miniconda3/envs/testenv/lib/python3.6/site-packages/numdifftools/core.py:297: in __call__
    results = self._derivative(xi, args, kwds)
../../../miniconda3/envs/testenv/lib/python3.6/site-packages/numdifftools/core.py:663: in _derivative_nonzero_order
    return self._apply_fd_rule(step_ratio, results, steps2)
../../../miniconda3/envs/testenv/lib/python3.6/site-packages/numdifftools/core.py:394: in _apply_fd_rule
    fd_rule = self._get_finite_difference_rule(step_ratio)
../../../miniconda3/envs/testenv/lib/python3.6/site-packages/numdifftools/core.py:378: in _get_finite_difference_rule
    fd_mat = self._fd_matrix(step_ratio, parity, num_terms)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
step_ratio = 2.0, parity = 1, nterms = 1
    @staticmethod
    def _fd_matrix(step_ratio, parity, nterms):
        """
        Return matrix for finite difference and complex step derivation.
    
        Parameters
        ----------
        step_ratio : real scalar
            ratio between steps in unequally spaced difference rule.
        parity : scalar, integer
            0 (one sided, all terms included but zeroth order)
            1 (only odd terms included)
            2 (only even terms included)
            3 (only every 4'th order terms included starting from order 2)
            4 (only every 4'th order terms included starting from order 4)
            5 (only every 4'th order terms included starting from order 1)
            6 (only every 4'th order terms included starting from order 3)
        nterms : scalar, integer
            number of terms
        """
        _assert(0 <= parity <= 6,
                'Parity must be 0, 1, 2, 3, 4, 5 or 6! ({0:d})'.format(parity))
        step = [1, 2, 2, 4, 4, 4, 4][parity]
        inv_sr = 1.0 / step_ratio
        offset = [1, 1, 2, 2, 4, 1, 3][parity]
        c0 = [1.0, 1.0, 1.0, 2.0, 24.0, 1.0, 6.0][parity]
        c = c0 / \
>           misc.factorial(np.arange(offset, step * nterms + offset, step))
E       AttributeError: module 'scipy.misc' has no attribute 'factorial'
../../../miniconda3/envs/testenv/lib/python3.6/site-packages/numdifftools/core.py:330: AttributeError
```

I do **not** see these on my laptop -- all the tests pass there.  This does not seem to have anything to do with my change to the `assert`, but what do I know?

ColCarroll replied:
my guess is a version of something (`numdifftools`?) changed under this diff. Comparing versions from the previous travis build and this one should let us know (and figure out if there's a good change). I'll dig into it after work.

rpgoldman replied:
That looks right -- I don't think it's anything I did. Is there some way to force a re-run of Travis on `master`, so that we can see if this failure also happens there (and my little tweak is innocent)?

rpgoldman replied:
@ColCarroll `numdifftools` didn't change, **but**  `scipy` 1.3.0 was released three days ago (17 May).  This looks like an upstream problem with `numdifftools`.  

And yes,`factorial` was removed from `scipy.misc`: https://github.com/scipy/scipy/commit/1cde30518fdc46c2c765beaa3e132e322ea7a579#diff-0763dd201fcbf9b088f9eb502b4f3d42

AFAICT this is fixed in the version of `numdifftools` on GitHub (or at least it is *different* -- factorial comes from `scipy.special` instead of `scipy.misc` now) *but* `numdifftools` has not seen a release in two years.

When you've got the factorial problem fixed, please squash merge this PR, or LMK if you would prefer me to smash it down to one commit for you to merge. Thanks!

ColCarroll reviewed (commented):
looks good to me once test pass! thanks for sticking with this @rpgoldman!

twiecki commented:
@rpgoldman What's your email?

rpgoldman commented:
Rebased onto the latest master.  If the tests pass now, I will squash this down to one commit so there's a clean history.

twiecki commented:
We can also just squash merge here.

On Tue, May 21, 2019, 17:31 rpgoldman <notifications@github.com> wrote:

> Rebased onto the latest master. If the tests pass now, I will squash this
> down to one commit so there's a clean history.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pymc-devs/pymc3/pull/3468?email_source=notifications&email_token=AAFETGHKZCZ72WPRHJ6W2ZLPWQIVHA5CNFSM4HLCBZAKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODV4JKNQ#issuecomment-494441782>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAFETGDGNESNL56V7UZ7TU3PWQIVHANCNFSM4HLCBZAA>
> .
>

rpgoldman committed:
Delete unused variable.

rpgoldman commented:
Pushed a cleaned up version of the history. Waiting for the checks, then I hope we can merge and close.

lucianopaz commented:
Thanks @rpgoldman!
